api/debug.go

```diff
@@ -22,77 +38,390 @@ type txTraceResult struct {
-	for i, h := range block.TransactionHashes {
+	args TransactionArgs,
+	blockNrOrHash rpc.BlockNumberOrHash,
+	config *tracers.TraceCallConfig,
+) (interface{}, error) {
+	tx, err := encodeTxFromArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Default address in case user does not provide one
+	from := d.config.Coinbase
+	if args.From != nil {
+		from = *args.From
+	}
+
+	var traceConfig *tracers.TraceConfig
+	if config != nil {
+		traceConfig = &config.TraceConfig
+	}
+
+	tracer, err := tracerForReceipt(traceConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	height, err := d.resolveBlockNumberOrHash(&blockNrOrHash)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := d.blocks.GetByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+
+	blocksProvider := replayer.NewBlocksProvider(
+		d.blocks,
+		d.config.FlowNetworkID,
+		tracer,
+	)
+	viewProvider := query.NewViewProvider(
+		d.config.FlowNetworkID,
+		flowEVM.StorageAccountAddress(d.config.FlowNetworkID),
+		d.registerStore,
+		blocksProvider,
+		120_000_000,
+	)
+
+	view, err := viewProvider.GetBlockView(block.Height)
+	if err != nil {
+		return nil, err
+	}
+
+	to := gethCommon.Address{}
+	if tx.To != nil {
+		to = *tx.To
+	}
+	opts := []query.DryCallOption{}
+	opts = append(opts, query.WithTracer(tracer))
+	if config.StateOverrides != nil {
+		for addr, overrideAccount := range *config.StateOverrides {
+			if overrideAccount.Nonce != nil {
+				opts = append(opts, query.WithStateOverrideNonce(addr, uint64(*overrideAccount.Nonce)))
+			}
+			if overrideAccount.Code != nil {
+				opts = append(opts, query.WithStateOverrideCode(addr, *overrideAccount.Code))
+			}
+			if overrideAccount.Balance != nil {
+				opts = append(opts, query.WithStateOverrideBalance(addr, (*big.Int)(*overrideAccount.Balance)))
+			}
+			if overrideAccount.State != nil {
+				opts = append(opts, query.WithStateOverrideState(addr, *overrideAccount.State))
+			}
+			if overrideAccount.StateDiff != nil {
+				opts = append(opts, query.WithStateOverrideStateDiff(addr, *overrideAccount.StateDiff))
+			}
+		}
+	}
+	_, err = view.DryCall(
+		from,
+		to,
+		tx.Data,
+		tx.Value,
+		tx.Gas,
+		opts...,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tracer.GetResult()
+}
+
+func (d *DebugAPI) executorAtBlock(block *models.Block) (*evm.BlockExecutor, error) {
+	snapshot, err := d.registerStore.GetSnapshotAt(block.Height)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get register snapshot at block height %d: %w", block.Height, err)
+	}
+	ledger := storage.NewRegisterDelta(snapshot)
+
+	return evm.NewBlockExecutor(
+		block,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
+}
 
-		txTrace, err := d.TraceTransaction(ctx, h, nil)
+func (d *DebugAPI) resolveBlockNumberOrHash(block *rpc.BlockNumberOrHash) (uint64, error) {
+	err := fmt.Errorf("%w: neither block number nor hash specified", errs.ErrInvalid)
```

Comment:
better to inline this error in `return`, so that we optimize for happy path where the err doesn't need to be created.

Reply:
Updated in https://github.com/onflow/flow-evm-gateway/pull/635/commits/c675e995bee0a128beec067c9f592be6745a711d.
Also moved some common utility functions to their own file, to make them re-usable.